File: sorting-and-searching/apartments/apartments.ts
function apartments(
  nApplicants: number,
  nApartments: number,
  maxAllowedDifference: number,
  desiredSizes: number[],
  availableSizes: number[]
): void {
  if (nApplicants < 1 || nApplicants > 200000) {
    console.log("number of applicants must be between 1 and 200000");
    return;
  }
  if (nApartments < 1 || nApartments > 200000) {
    console.log("number of apartments must be between 1 and 200000");
    return;
  }
  if (maxAllowedDifference < 0 || maxAllowedDifference > 1000000000) {
    console.log("maximum allowed difference must be between 0 and 1e9");
    return;
  }
  if (nApplicants !== desiredSizes.length) {
    console.log("number of applicants must match length of desired sizes");
    return;
  }
  if (nApartments !== availableSizes.length) {
    console.log("number of apartments must match length of available sizes");
    return;
  }

  desiredSizes.sort((a, b) => a - b);
  availableSizes.sort((a, b) => a - b);

  let i = 0;
  let j = 0;
  let success = 0;

  while (i < nApplicants && j < nApartments) {
    if (Math.abs(desiredSizes[i] - availableSizes[j]) <= maxAllowedDifference) {
      success++;
      i++;
      j++;
    } else if (availableSizes[j] < desiredSizes[i] - maxAllowedDifference) {
      j++;
    } else {
      i++;
    }
  }

  console.log(success);
}

apartments(4, 3, 5, [60, 45, 80, 60], [30, 60, 75]);

File: sorting-and-searching/apartments/apartments.py
def apartments(
    n_applicants: int,
    n_apartments: int,
    k: int,
    desired_size: list[int],
    available_size: list[int],
) -> None:

    if not (1 <= n_applicants <= 200_000):
        print("Number of applicants must be between 1 and 200000")
        return
    if not (1 <= n_apartments <= 200_000):
        print("Number of apartments must be between 1 and 200000")
        return
    if not (0 <= k <= 1_000_000_000):
        print("Maximum allowed difference must be between 0 and 1e9")
        return
    if len(desired_size) != n_applicants:
        print("Length of desired_size list must match n_applicants")
        return
    if len(available_size) != n_apartments:
        print("Length of available_size list must match n_apartments")
        return

    for idx, size in enumerate(desired_size):
        if not (1 <= size <= 1_000_000_000):
            print(f"desired_size[{idx}] must be between 1 and 1e9")
            return

    for idx, size in enumerate(available_size):
        if not (1 <= size <= 1_000_000_000):
            print(f"available_size[{idx}] must be between 1 and 1e9")
            return

    desired_size.sort()
    available_size.sort()

    i = j = 0
    success = 0

    while i < n_applicants and j < n_apartments:
        if abs(desired_size[i] - available_size[j]) <= k:
            success += 1
            i += 1
            j += 1
        elif available_size[j] < desired_size[i] - k:
            j += 1
        else:
            i += 1

    print(success)


apartments(4, 3, 5, [60, 45, 80, 60], [30, 60, 75])
